### DISCUSSION TITLE
Popup close behavior

### DISCUSSION THREAD

**Discussion by progsource** [godotengine/godot-proposals#5135] [GUI]
Based on https://github.com/godotengine/godot/blob/1b06b668c9f0c42927a427f9c9057711a4adafe4/scene/gui/popup.cpp#L37-L42

```cpp
void Popup::_input_from_window(const Ref<InputEvent> &p_event) {
	Ref<InputEventKey> key = p_event;
	if (key.is_valid() && key->is_pressed() && key->get_keycode() == Key::ESCAPE) {
		_close_pressed();
	}
}
```

closing of a popup always happens on escape.
I would like it more if a programmer could decide if that is actually desired behavior, instead of forcing it there - or at least decide on the key code.

Also when the game window loses focus, it will close popups

https://github.com/godotengine/godot/blob/cf95056c918b30509c3b181d672e6cfdc971d34f/scene/gui/popup.cpp#L94-L95

```cpp
		case NOTIFICATION_APPLICATION_FOCUS_OUT: {
			_close_pressed();
```

I'm not so sure if that is actually always desired behavior in-game.

What do you guys think?

**Comment by progsource** [comment]
related issue: https://github.com/godotengine/godot/issues/67005

**Comment by bruvzg** [comment]
You can always use `Window` instead of `Popup` if you want to customize it. `Popup` is specifically designed to have auto-close behavior, it's basically its only difference for the `Window`.

**Reply by bruvzg** [reply]
To be more specific, if you want a customizable child window, it's done by setting `transient` flag of the `Window`.

**Comment by hawkerm** [comment]
None of this was intuitive at all to understand the differences between popups and windows in the docs and the nuances between them. Looks like I need to be using a `Window` as well...
